### ISSUE TITLE
System request: Chinese Yawei machine

### ISSUE THREAD

**Issue by morinted** [openstenoproject/plover#749]
Devin Tankersly writes:

> I'm an American currently doing an MA in linguistics in Taiwan, and I've long felt plagued by the (slow) speed at which I type in Chinese. I came across Plover not that long ago, and after a few false starts, I have recently been able to get into the swing of learning to type with it. 
> 
> Soon after getting into the groove of learning, I started to wonder how stenography could work for Chinese. By looking through some of the older posts here and around some other sites, I learned of the existence of the Yawei stenography machine (亞偉速錄機), and I went ahead and ordered a textbook for it, since I was super curious how it was all set up, and the details I could find online were quite sparse.
> 
> Here's what the layout looks like:
> 
> ```
> A N I G D  D G I N A
> O E U W Z  Z W U E O
>       B X  X B
> ``` 
>
> The layout is largely similar to Plover's, with the keys of course being renamed, and also doing away with the * key(s) and # bar. Notably, the layout is mirrored down the middle, so that the left and right hands each have access to the same keys -- as each side governs one syllable, the typist can easily type two syllables at a time, with the left hand governing the first syllable, and the right governing the second. This is quite handy for a number of reasons: first, Mandarin Chinese quite likes grouping morphemes into disyllabic groups (students of Mandarin learn these largely as compounds); second, this helps deal with much of the problems around homophones, for example a syllable like 'fu', stroked as XBU (Yawei does not consider tone), could be 夫 fū "man/husband" 服 fú "serve/clothes" 府 fǔ "seat of government" 婦 fù "(married) woman", among a variety of other possibilities, but the disyllabic 'fufu' (stroked XBU:XBU) is most likely 夫婦 fūfù "husband and wife", with just a few other (rather rare) possibilities. 
> 
> After receiving my Yawei textbook in the mail, I excitedly went through the opening material and the first chapter. I had hoped to kind of "hack" Plover into being able to use this method, largely by using the existing layout but just pretending as if the keys are as they should be on a Yawei machine, and adding phrases to the user dictionary as I go. But I quickly realized that this is simply not feasible due to the nature of the default steno layout in Plover, as the left hand would be lacking three keys (i.e. Yawei splits Plover's left S into A and O, and Yawei's left D and Z are subsumed under Plover's *). 
> 
> I looked around a bit to see what other users had done for adapting Plover to other languages, and I am fairly certain that someone more coding-savvy than I am could modify the layout such that Yawei could be implemented. Were that to happen, I would happily do my part to fill out a dictionary and possibly even make some practice tools for learning the system, if it seems like people would be interested in that. I am not sure about the copyright situation with something like this, but because Plover seems to be based in some part on existing theories, I hope it would be deemed appropriate. That being said, after a more thorough searching, I've found a website that contains much of the same information as the textbook, so it seems fair to say that this information is available to the public: http://ptr.chaoxing.com/nodedetailcontroller/visitnodedetail?knowledgeId=2183315
> 
> As a side note, I have considered buying a Yawei steno machine, but I am hesitant to do so for two main reasons: such a machine will likely only be able to type simplified Chinese, and I have my doubts about how easy it would be to modify the corresponding software to allow for traditional characters; as an MA student without full time employment, I do not have a lot of disposable income to spend on something that might not be able to type what I would need it to.
> 
> tl;dr: if someone could adapt Plover's layout to fit Yawei's (as shown above), I would super appreciate it and would work on writing up the dictionary for Chinese steno on Plover.

**Comment by morinted** [comment]
Last time I looked into Chinese steno (I believe it was someone on Discord who was talking about it) we quickly ran into the issue that the software prompts the user to select between several alternatives when ambiguous entries are entered. Is that a trait of the Yawei system?

**Comment by lokivari** [comment]
I can't confirm what the software does in the case of ambiguous input, but after skimming through most of the textbook that I got, it seems like Yawei has other strategies for dealing with homophones. In addition to the emphasis on disyllabic chunking, the W and X keys are used as disambiguators, much like * in Plover. For example, the syllable 'you' (stroked IEO) could be 有 yǒu "have", 由 yóu "from/by", or 又 yòu "again", and since in many cases they would not fit easily into two syllable phrases, they might need to be typed alone, so 有 is stroked X:IEO, 由 is W:IEO, and 又 is one of the few cases where both disambiguators get used, so it is stroked XW:IEO.

I should mention though that most input methods for Chinese do deal with ambiguous input in the way that you described, where the user has to choose between a few options.

**Comment by morinted** [comment]
That sounds promising. I've got a scattered list of questions:

- Is `:` used as a stroke delimiter or a side indicator? E.g. in English steno we do `STOP/-T/KAT` (stop the cat, the dash is used to indicate the side that the T is on)
- Is there a sample dictionary available, or are you expected to build your own?
- The Chinese steno machines are much cheaper than the North American ones, right? If I wanted to purchase one with international shipping, do you know of a good supplier?
- Does the theory call itself a realtime theory (it's always a big deal in the English books as to whether you are realtime or not)

**Comment by lokivari** [comment]
I probably should have clarified about it before, but according to Yawei convention, the colon (`:`) is used to specify which side the keys are on, similar to the hyphen (`-`) in English steno. I'll try to use the English steno notation from now on for the sake of clarity. Also, since both sides of the keyboard have the same keys, the convention is to write the code in the same order, i.e. there is a set order for the keys, regardless of whether it is meant to appear on the left or right hand side. Looking again at the 'fufu' example I used before, you'll see that the code is written as `XBU-XBU`, even though if we were to read the keys from left to right it "should" be `UBX-XBU`. I think it's done this way a) for ease of reading, and b) because it's the same code, just appearing on one side or the other. The only place where Yawei convention reorders the keys for ease of reading (I'll list the standard order below) is `NA` -> `AN`, as this combination represents the sound 'an', but this convention doesn't seem to be followed by all sources, so I won't use it here.

[Standard Yawei key order] 
`X B D Z G W I U N E A O`

After a more careful search through the textbook and some online resources, including the Yawei company website, I have confirmed that it does indeed use the number/function keys above the keyboard to select the output for ambiguous input. This has to do with the fact that some two syllable combinations do correspond to a large number of phrases, especially if tone is not considered. To give an example, 'shijian' (stroked as `XZ-GINA`) corresponds to a few common (and not so common) compounds, and I've listed some below in rough order of frequency:

時間 shíjiān "time"
實踐 shíjiàn "(put into) practice"
事件 shìjiàn "incident"
始建 shǐjiàn "start building"
失檢 shījiǎn "(be) indiscreet"
屍檢 shījiǎn "autopsy"
時艱 shíjiān "times of trouble, crisis"
識見 shìjiàn "knowledge and experience"

With the exception of 時間 shíjiān "time", which has its own brief in `XZ-X`, all the other above combinations would be stroked in the same way, and barring any other contextual disambiguation (e.g. a set phrase could be in the dictionary so that XZ-GINA would resolve to 事件 if preceded by 緊急 jǐnjí "urgent", so that the whole phrase would be 緊急事件 jǐnjíshìjiàn "emergency", stroked as `GIN-GI/XZ-GINA`), the only way to solve the ambiguity would be through forced selection. On the Yawei machine, this is done through the number/function keys above the keyboard, and since there might be a high number of possibilities, the user may have to switch to the next "page" of options. A hypothetical example using the compounds above would assign each combination a page/index number, as given below.

實踐 shíjiàn "(put into) practice" (1-1)
事件 shìjiàn "incident" (1-2)
始建 shǐjiàn "start building" (1-3)
失檢 shījiǎn "(be) indiscreet" (1-4)
屍檢 shījiǎn "autopsy" (1-5)
時艱 shíjiān "times of trouble, crisis" (2-1)
識見 shìjiàn "knowledge and experience" (2-2)

So in my hypothetical example, to type 識見 shìjiàn "knowledge and experience", the user would stroke `XZ-GINA` for 'shijian' and then hit another key to go to the next page of outputs, and hit 2 to output 識見.

I think that the problem this sort of input-output system poses for Plover is the whole "forced selection" thing, where the user is given a set of choices to make. I can foresee a work-around for this, where the number keys are used to fulfill this role, so that the dictionary would have 實踐 shíjiàn "(put into) practice" as `XZ-GINA/1`, 事件 shìjiàn "incident" as `XZ-GINA/2`, and so on. If a particular syllable combination has more than let's say 9 possibilities, the number keys could also be chorded, so the tenth output could be chosen using a following `/12`, the next one would be `/13`, and so on.

The drawback to this strategy (besides the fingers having to leave home position) seems to be the rote memorization required, especially when the choices are not made visible to the user. I don't know how feasible it would be, but if the "suggestions" window could be made to display the choices in much the same way as the Yawei machine does, this could help ease typists' learning of the arbitrary number assignment.

This is neither here nor there, but I find it interesting that another strategy in Yawei for dealing with typing individual characters (which, given the homophony issue, is rather difficult if one can only use the syllable by itself), is to input an unambiguous stroke and then delete either the first or second character. For example, to type 至 zhì "arrive, to, until, etc" the textbook has students stroke `Z-IU` for 至於 zhìyú "as for, as to (some topic)", and then delete 於 yú "in, at, on" (a kind of general purpose preposition) using `-W` (`W-W` deletes the preceding stroke instead of just one character). Granted, I don't know how often professional stenographers rely on this sort of strategy, but based on what's in the textbook and in the software package from the Yawei company website, it seems to be a standard approach.

When dealing with personal names, which may use rare characters in somewhat unpredictable combinations, I can see the value to this strategy, especially given the fact that many people when explaining how to write their name give example compounds to the listener, e.g. my Chinese name has the somewhat uncommon 駿 jùn "steed, spirited horse", and so to make sure people don't misinterpret my name as having the homophonous and more frequently used 俊 jùn "handsome", I always say that it is 駿馬的駿 jùnmǎ de jùn "the 'jun' in 'junma' (fine horse)".

Regarding the dictionary, I had originally envisioned typing one up as I went through the textbook, but considering the vast number of characters and character combinations required, this does not seem super reasonable. The [Yawei website has some free software](http://www.suluji.com/index.php/Service/download) that could alleviate much of this problem, but I am not sure how to extract that information. Barring that, considering the one-to-one correspondence of syllable structure to Yawei input code, it would not be impossible to take a text file containing the pinyin for each character / compound and have the pinyin swapped out for Yawei code, and I can think of a few sources for tables like that.

As to whether or not the theory is considered "realtime", I guess I would have to ask what "realtime" means in this case. If it's meant to say that the theory allows stenographers to type in realtime, catching what people say as they say them, then Yawei would qualify, as it claims to allow stenographers to type at the speed of speech or faster.

Going back through the other conversations around Chinese steno on the Google group, I found someone posted [a seller via Taobao](https://yaoyaosuji.world.taobao.com/shop/view_shop.htm) (the Chinese equivalent to Amazon/ebay) that has a good number of [new](https://world.taobao.com/item/8625760517.htm?fromSite=main&spm=a312a.7700824.w1003-947942690.7.h7Tw0g&scene=taobao_shop) and [used](https://world.taobao.com/item/8631500335.htm?fromSite=main&spm=a312a.7700824.w1017-947942784.14.1HCeB5&) Yawei steno machines for under $500 USD. I ordered my textbook through Taobao, and I've generally found it to be a reputable service, but I haven't bought anything all that expensive from the site before.

Sorry for that huge wall of text! If it seems like it would be useful, sometime over the next week or so I can write up a brief overview of the system so that all of this information is better organized and more easily accessible.

**Comment by mkrnr** [comment]
Closing this since it would be implemented as a plugin.
